File: app/src/main/java/com/zygne/confetti/engine/view/ExplosionSurface.java
package com.zygne.confetti.engine.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import com.zygne.confetti.engine.explosions.BaseExplosion;
import com.zygne.confetti.engine.explosions.Emitter;
import com.zygne.confetti.engine.explosions.Explosion;
import com.zygne.confetti.engine.physics.Physics;
import com.zygne.confetti.engine.util.FpsCounter;

/**
 * Created by Bardur Thomsen on 9/17/18.
 */
public class ExplosionSurface extends SurfaceView implements SurfaceHolder.Callback, Runnable {

    public static final String TAG = ExplosionSurface.class.getSimpleName();

    public static final int DEFALUT_PARTICLE_NUMBER = 64;

    public Callback callback;
    private SurfaceHolder surfaceHolder;
    private Thread thread = null;
    private boolean running = false;
    private int screenWidth = 0;
    private int screenHeight = 0;
    private BaseExplosion explosion;
    private int particles;
    private FpsCounter fpsCounter = new FpsCounter();
    private int timeToNotify;


    public ExplosionSurface(Context context) {
        super(context);

        setFocusable(true);

        // Get SurfaceHolder object.
        surfaceHolder = null;
        surfaceHolder = this.getHolder();
        // Add current object as the callback listener.
        surfaceHolder.addCallback(this);

        // Set the SurfaceView object at the top of View object.
        setZOrderOnTop(true);

        setBackgroundColor(Color.TRANSPARENT);

        particles = DEFALUT_PARTICLE_NUMBER;

    }

    public ExplosionSurface(Context context, int numberOfParticles) {
        super(context);

        setFocusable(true);

        // Get SurfaceHolder object.
        surfaceHolder = null;
        surfaceHolder = this.getHolder();
        // Add current object as the callback listener.
        surfaceHolder.addCallback(this);

        // Set the SurfaceView object at the top of View object.
        setZOrderOnTop(true);

        setBackgroundColor(Color.TRANSPARENT);

        particles = numberOfParticles;
    }

    @Override
    public void surfaceCreated(SurfaceHolder surfaceHolder) {

        // Create the child thread when SurfaceView is created.
        thread = new Thread(this);
        // Start to run the child thread.
        thread.start();
        // Set thread running flag to true.
        running = true;

        // Get screen width and height.
        screenHeight = getHeight();
        screenWidth = getWidth();

        explosion = new Explosion(screenWidth / 2, screenHeight / 2, particles);
        explosion.setPhysics(new Physics(0f, 0f, 1f));

        Log.d(TAG, explosion.toString());
    }

    @Override
    public void surfaceChanged(SurfaceHolder surfaceHolder, int i, int i1, int i2) {
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder surfaceHolder) {
        // Set thread running flag to false when Surface is destroyed.
        // Then the thread will jump out the while loop and complete.
        running = false;
    }


    @Override
    public void run() {
        while (running) {

            Canvas canvas= null;
            try {
                // Get Canvas from Holder and lock it.
                canvas = this.surfaceHolder.lockCanvas();

                canvas.drawColor(0, android.graphics.PorterDuff.Mode.CLEAR);
                explosion.update(0);
                explosion.render(canvas);

            }catch(Exception e)  {
                // Do nothing.
            } finally {
                if(canvas!= null)  {
                    // Unlock Canvas.
                    this.surfaceHolder.unlockCanvasAndPost(canvas);
                }
            }

            fpsCounter.countFrames();
            timeToNotify++;

            if(timeToNotify > 200){
                if(callback != null){
                    callback.onFpsUpdate(fpsCounter.getFps());
                }
            }
        }
    }

    public void setCallback(Callback callback){
        this.callback = callback;
    }

    public void resetExplosion(int particles){
        this.particles = particles;
        explosion = new Explosion(screenWidth / 2, screenHeight / 2, particles);
        explosion.setPhysics(new Physics(0f, 0f, 1f));
    }

    public void resetExplosion(int particles, int type){
        this.particles = particles;

        if(type == 1) {
            explosion = new Explosion(screenWidth / 2, screenHeight / 2, particles);
        } else {
            explosion = new Emitter(screenWidth / 2, screenHeight / 2, particles);
        }

        explosion.setPhysics(new Physics(0f, 0f, 1f));
    }

    public void updateWind(float wind){
        explosion.getPhysics().setWind(wind);
    }

    public void updateGravity(float gravity){
        explosion.getPhysics().setGravity(gravity);
    }

    public interface Callback{
        void onFpsUpdate(int fps);
    }

}

File: app/src/main/java/com/zygne/confetti/engine/view/ExplosionSurface.kt
package com.zygne.confetti.engine.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.PorterDuff
import android.util.AttributeSet
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import com.zygne.confetti.engine.explosions.BaseExplosion
import com.zygne.confetti.engine.explosions.Emitter
import com.zygne.confetti.engine.explosions.Explosion
import com.zygne.confetti.engine.explosions.Fire
import com.zygne.confetti.engine.particles.BaseParticle
import com.zygne.confetti.engine.physics.Physics
import com.zygne.confetti.engine.util.FpsCounter

class ExplosionSurface(context: Context?, attributeSet: AttributeSet?) :
    SurfaceView(context, attributeSet), SurfaceHolder.Callback, Runnable {
    private var callback: Callback? = null
    private var surfaceHolder: SurfaceHolder?
    private var thread: Thread? = null
    private var running = false
    private var screenWidth = 0
    private var screenHeight = 0
    private var explosion: BaseExplosion? = null
    private var particles: Int
    private val fpsCounter = FpsCounter()
    private var timeToNotify = 0

    init {
        isFocusable = true
        surfaceHolder = null
        surfaceHolder = this.holder
        surfaceHolder!!.addCallback(this)
        setZOrderOnTop(false)
        setBackgroundColor(Color.TRANSPARENT)
        particles = DEFAULT_PARTICLE_NUMBER
    }

    override fun surfaceCreated(surfaceHolder: SurfaceHolder) { // Create the child thread when SurfaceView is created.
        thread = Thread(this)
        thread!!.start()
        running = true
        screenHeight = height
        screenWidth = width
        explosion = Explosion(
            (screenWidth / 2).toFloat(),
            (screenHeight / 2).toFloat(),
            particles,
            BaseParticle.DEFAULT_MAX_SPEED
        )
        explosion!!.physics = Physics(0f, 0f, 1f)
        Log.d(TAG, explosion.toString())
    }

    override fun surfaceChanged(surfaceHolder: SurfaceHolder, i: Int, i1: Int, i2: Int) {}
    override fun surfaceDestroyed(surfaceHolder: SurfaceHolder) {
        running = false
    }

    override fun run() {
        while (running) {
            var canvas: Canvas? = null
            try {
                canvas = surfaceHolder!!.lockCanvas()
                canvas.drawColor(0, PorterDuff.Mode.CLEAR)
                explosion!!.update(0f)
                explosion!!.render(canvas)
            } catch (e: Exception) {
            } finally {
                if (canvas != null) {
                    surfaceHolder!!.unlockCanvasAndPost(canvas)
                }
            }
            fpsCounter.countFrames()
            timeToNotify++
            if (timeToNotify > 200) {
                if (callback != null) {
                    callback!!.onFpsUpdate(fpsCounter.fps)
                }
            }
        }
    }

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    fun resetExplosion(particles: Int, type: Int, speed: Int) {
        this.particles = particles
        explosion = if (type == 1) {
            Explosion((screenWidth / 2).toFloat(), (screenHeight / 2).toFloat(), particles, speed)
        } else {
            Emitter((screenWidth / 2).toFloat(), (screenHeight / 2).toFloat(), particles, speed)
        }
        explosion!!.physics = Physics(0f, 0f, 1f)
    }

    fun updateWind(wind: Float) {
        explosion!!.physics.wind = wind
    }

    fun updateGravity(gravity: Float) {
        explosion!!.physics.gravity = gravity
    }

    interface Callback {
        fun onFpsUpdate(fps: Int)
    }

    companion object {
        const val TAG = "ExplosionSurface"
        const val DEFAULT_PARTICLE_NUMBER = 64
    }
}